test/cpp/end2end/xds/xds_fallback_end2end_test.cc

```diff
@@ -63,40 +64,65 @@ class XdsFallbackTest : public XdsEnd2endTest {
+                                absl::string_view target = "") {
     Listener listener = default_listener_;
     RouteConfiguration route_config = default_route_config_;
     Cluster cluster = default_cluster_;
-    // Server 0 uses default resources when no authority, to enable using more
+    // Target 0 uses default resources when no authority, to enable using more
     // test framework functions
-    if (target_id > 0 && authority.empty()) {
-      listener.set_name(absl::StrFormat("server%d.example.com", target_id));
-      cluster.set_name(absl::StrFormat("cluster%d", target_id));
+    if (!target.empty()) {
+      listener.set_name(target);
+      cluster.set_name(absl::StrFormat("cluster_%s", target));
       cluster.mutable_eds_cluster_config()->set_service_name(
-          absl::StrFormat("eds%d", target_id));
-      route_config.set_name(absl::StrFormat("route%d", target_id));
+          absl::StrFormat("eds_%s", target));
+      route_config.set_name(absl::StrFormat("route_%s", target));
       route_config.mutable_virtual_hosts(0)
           ->mutable_routes(0)
           ->mutable_route()
           ->set_cluster(cluster.name());
-    } else if (!authority.empty()) {
     }
     SetListenerAndRouteConfiguration(balancer, listener, route_config);
     balancer->ads_service()->SetCdsResource(cluster);
     balancer->ads_service()->SetEdsResource(BuildEdsResource(
         EdsResourceArgs(
             {{"locality0", CreateEndpointsForBackends(backend, backend + 1)}}),
         cluster.eds_cluster_config().service_name()));
-    return listener.name();
+  }
+
+  void ConfigureAuthority(const std::string& eds_server_name,
+                          const std::string& cluster_name,
+                          const std::string& route_config_name,
+                          const std::string& listener_name,
+                          const EdsResourceArgs& args) {
+    fallback_balancer_->ads_service()->SetEdsResource(
+        BuildEdsResource(args, eds_server_name));
+    // New cluster
+    Cluster new_cluster = default_cluster_;
+    new_cluster.set_name(cluster_name);
+    new_cluster.mutable_eds_cluster_config()->set_service_name(eds_server_name);
+    fallback_balancer_->ads_service()->SetCdsResource(new_cluster);
+    // New Route
+    RouteConfiguration new_route_config = default_route_config_;
+    new_route_config.set_name(route_config_name);
+    new_route_config.mutable_virtual_hosts(0)
+        ->mutable_routes(0)
+        ->mutable_route()
+        ->set_cluster(cluster_name);
+    // New Listener
+    Listener listener = default_listener_;
+    listener.set_name(listener_name);
+    SetListenerAndRouteConfiguration(fallback_balancer_.get(), listener,
+                                     new_route_config);
   }
 
  protected:
   std::unique_ptr<BalancerServerThread> fallback_balancer_;
 };
 
 TEST_P(XdsFallbackTest, FallbackAndRecover) {
+  grpc_core::testing::ScopedEnvVar fallback_enabled(
```

Comment:
Please add a test that shows that fallback does *not* work if the env var is unset.